### PR TITLE
use bump2version

### PR DESCRIPTION
**Pull request by ThomasWaldmann**
this tool can be used to automatically update the version numbers in misc. files (currently we have 3 places), so that this does not have to be done manually.

also it adds a corresponding git tag and, as configured now, signs the tag.

this avoids mistakes like forgetting some places, doing inconsistent updates or forgetting to set a tag.

### CODE REVIEW

**Comment by ThomasWaldmann** [comment]
TODO: refactor the code so it derives all required values (see TODO in the code) from the VERSION_STR.

**Comment by rexfue** [comment]
Can we then use an extra version.h file with the version number? I want to use the version in my GUI, an if its an extra h-file I can easily include it.

**Comment by rexfue** [comment]
OK, changes on VERSION_STR can be made, no problem with that.

File: multigeiger/multigeiger.ino [hunk 1]
```diff
@@ -754,25 +751,21 @@ void DisplayStartscreen(void) {
 
   u8x8.clear();
```
Comment (ThomasWaldmann) [inline multigeiger/multigeiger.ino]: the above line clears the display.

File: multigeiger/multigeiger.ino [hunk 2]
```diff
@@ -754,25 +751,21 @@ void DisplayStartscreen(void) {
...
   #if CPU == STICK
-  // Display is only 4 lines by 8 characters; lines counting from 2 to 5
-  u8x8.setFont(u8x8_font_5x8_f);        // use really small font
-  for (int i = 2; i < 6; i++) {
-    u8x8.drawString(0, i, "        ");  // clear all 4 lines
-  }
```
Comment (ThomasWaldmann) [inline multigeiger/multigeiger.ino]: this cleared the display again!? removed that.

File: multigeiger/multigeiger.ino [hunk 3]
```diff
@@ -754,25 +751,21 @@ void DisplayStartscreen(void) {
...
+  // Display is only 4 lines by 8 characters; lines counting from 2 to 5.
+  snprintf(line, 9, "%s", VERSION_STR);  // 8 chars + \0 termination
+  u8x8.setFont(u8x8_font_5x8_f);
   u8x8.drawString(0, 2, "Geiger-");
   u8x8.drawString(0, 3, " Counter");
   u8x8.drawString(0, 4, "Version:");
-  sprintf(line, "%d.%d.%d", VERSION_MAJOR, VERSION_MINOR, VERSION_PATCH);
   u8x8.drawString(0, 5, line);
   #else
+  // Display is only H lines by W characters; lines counting from A to B. -- TODO
```
Comment (ThomasWaldmann) [inline multigeiger/multigeiger.ino]: @rexfue do you know or do we need to find out?

File: multigeiger/multigeiger.ino [hunk 4]
```diff
@@ -754,25 +751,21 @@ void DisplayStartscreen(void) {
 
   u8x8.clear();
   #if CPU == STICK
-  // Display is only 4 lines by 8 characters; lines counting from 2 to 5
-  u8x8.setFont(u8x8_font_5x8_f);        // use really small font
-  for (int i = 2; i < 6; i++) {
-    u8x8.drawString(0, i, "        ");  // clear all 4 lines
-  }
+  // Display is only 4 lines by 8 characters; lines counting from 2 to 5.
```
Comment (ThomasWaldmann) [inline multigeiger/multigeiger.ino]: the docs say X and Y both start from 0.

so, the docs are wrong? or do we want to leave 2 empty lines at the top?

File: multigeiger/multigeiger.ino [hunk 5]
```diff
@@ -754,25 +751,21 @@ void DisplayStartscreen(void) {
...
+  snprintf(line, 9, "%s", VERSION_STR);  // 8 chars + \0 termination
+  u8x8.setFont(u8x8_font_5x8_f);
   u8x8.drawString(0, 2, "Geiger-");
   u8x8.drawString(0, 3, " Counter");
   u8x8.drawString(0, 4, "Version:");
-  sprintf(line, "%d.%d.%d", VERSION_MAJOR, VERSION_MINOR, VERSION_PATCH);
   u8x8.drawString(0, 5, line);
   #else
+  // Display is only H lines by W characters; lines counting from A to B. -- TODO
+  snprintf(line, 15, "%s", VERSION_STR);  // 14 chars + \0 termination
```
Comment (ThomasWaldmann) [inline multigeiger/multigeiger.ino]: the stuff below is 14 chars wide, so I assumed this is the max. for the version, too.

File: multigeiger/multigeiger.ino [hunk 6]
```diff
@@ -754,25 +751,21 @@ void DisplayStartscreen(void) {
...
-  sprintf(line, "V%d.%d.%d", VERSION_MAJOR, VERSION_MINOR, VERSION_PATCH);
-  u8x8.setCursor(7 - strlen(line) / 2, 4);
-  u8x8.print(line);
-  u8x8.setCursor(1, 6);
-  u8x8.print("Info:boehri.de");
+  u8x8.drawString(0, 2, "Geiger-Counter");
+  u8x8.drawString(0, 3, "==============");
+  u8x8.drawString(0, 4, line);
+  u8x8.drawString(0, 6, "Info:boehri.de");
   #endif
```
Comment (ThomasWaldmann) [inline multigeiger/multigeiger.ino]: be consistent with the code for the small display: use drawString for the bigger display also.

File: multigeiger/multigeiger.ino [hunk 7]
```diff
@@ -419,13 +417,12 @@ void setup() {
   // just for fun
   log(INFO, "Let's go!");
 
-  // build revString
-  sprintf(revString, "V%d.%d.%d %s", VERSION_MAJOR, VERSION_MINOR, VERSION_PATCH, VERSION_DATE);
```
Comment (ThomasWaldmann) [inline multigeiger/multigeiger.ino]: will now only be the full VERSION_STR (including "-dev" for development versions).

**Comment by ThomasWaldmann** [comment]
TODO:
- practically test initial OLED display output on WIFI and STICK devices.
- check version on lora data (should be ok, but better check)
- check serial output (should be ok also)

**Comment by rexfue** [comment]
checked OLEDs - see comments above
checked LoRa - OK
checked serial - OK

On the small display we should omit the "V" and replace it with a space. If one day we have versions above 9, there will be no more space.

**Comment by ThomasWaldmann** [comment]
I don't see any "comments above". Did you submit them?

Having a major release usually means breaking / major changes, these do not happen too often, so reaching major release 10 will likely take 10-20 years for us, if ever.

So, for the simplicity of the code, we'll solve that when/if we reach 10 and keep the V for now. :-)

**Comment by rexfue** [comment]
Sorry - forget to submit.

Agreed on the "V" - by the I may no longer be here ...

File: multigeiger/multigeiger.ino [hunk 8]
```diff
@@ -754,25 +744,21 @@ void DisplayStartscreen(void) {
...
+  snprintf(line, 15, "%s", VERSION_STR);  // 14 chars + \0 termination
   u8x8.setFont(u8x8_font_7x14_1x2_f);
-  u8x8.println("Geiger-Counter");
-  u8x8.println("==============");
-  sprintf(line, "V%d.%d.%d", VERSION_MAJOR, VERSION_MINOR, VERSION_PATCH);
-  u8x8.setCursor(7 - strlen(line) / 2, 4);
-  u8x8.print(line);
-  u8x8.setCursor(1, 6);
-  u8x8.print("Info:boehri.de");
+  u8x8.drawString(0, 2, "Geiger-Counter");
```
[the remark this answers is not in the document]
Comment (ThomasWaldmann) [reply multigeiger/multigeiger.ino]: ok, i'll check/fix.

[the remark this answers is not in the document]
Comment (ThomasWaldmann) [reply multigeiger/multigeiger.ino]:
```
  u8x8.drawString(0, 2, "Geiger-Counter");
  u8x8.drawString(0, 3, "==============");
  u8x8.drawString(0, 4, line);
  u8x8.drawString(0, 6, "Info:boehri.de");
```

0,2,4,6 can't be correct - I guess we want to have the underlining directly under the "Geiger-Counter" string.

1,2,4,6 maybe?

File: multigeiger/multigeiger.ino (hunk 1 above)
Comment (ThomasWaldmann) [reply multigeiger/multigeiger.ino]: Isn't the library expected to be generic for all supported displays?

So, if it doesn't clear, file a bug against the library?

Comment (ThomasWaldmann) [reply multigeiger/multigeiger.ino]: And does it still happen with the latest lib release?

File: multigeiger/multigeiger.ino (hunk 3 above)
Comment (ThomasWaldmann) [reply multigeiger/multigeiger.ino]: And the result for H and W is ... ?

File: multigeiger/multigeiger.ino (hunk 4 above)
Comment (ThomasWaldmann) [reply multigeiger/multigeiger.ino]: Again: file a bug against the library? It's docs say 0/0.

File: multigeiger/multigeiger.ino (hunk 5 above)
Comment (ThomasWaldmann) [reply multigeiger/multigeiger.ino]: OK, H=4, W=16.

File: multigeiger/multigeiger.ino (hunk 8 above)
[the remark this answers is not in the document]
Comment (ThomasWaldmann) [reply multigeiger/multigeiger.ino]: Ehrm, above you said H is 4.

So does line 0 and 1 overlap or why do you say 0, 2, 4, 6 (normally that would mean H==7)?

File: multigeiger/multigeiger.ino [hunk 9]
```diff
@@ -384,8 +375,17 @@ char *buildSSID() {
...
   u8x8.begin();
+  // On the WIFI CPU u8x8.begin() cleared the display.
+  #if CPU == STICK
+  // But not on the STICK CPU, so we do it here.
+  // Display is only 4 lines by 8 characters; lines counting from 2 to 5
+  u8x8.setFont(u8x8_font_5x8_f);        // use really small font
+  for (int i = 2; i < 6; i++) {
+    u8x8.drawString(0, i, "        ");  // clear all 4 lines
+  }
+  #endif
```
Comment (ThomasWaldmann) [inline multigeiger/multigeiger.ino]: if u8x8.begin() clears the WIFI display, but not the STICK display, doing both here should be structurally better, right?

File: multigeiger/multigeiger.ino [hunk 10]
```diff
@@ -419,13 +419,12 @@ void setup() {
...
   log(INFO, "Let's go!");
 
-  // build revString
-  sprintf(revString, "V%d.%d.%d %s", VERSION_MAJOR, VERSION_MINOR, VERSION_PATCH, VERSION_DATE);
-
   #if SEND2LORA
-  // build LoRa software version
-  lora_software_version = (VERSION_MAJOR << 12) + (VERSION_MINOR << 4) + VERSION_PATCH;
+  int major, minor, patch;
+  sscanf(VERSION_STR, "V%d.%d.%d", &major, &minor, &patch);
```
Comment (ThomasWaldmann) [inline multigeiger/multigeiger.ino]: why is major, minor coloured differently here than patch?

File: multigeiger/multigeiger.ino [hunk 11]
```diff
@@ -751,28 +750,23 @@ int gen_charge_pulses(int max_charge_pulses) {
...
+  snprintf(line, 9, "%s", VERSION_STR);  // 8 chars + \0 termination
+  u8x8.setFont(u8x8_font_5x8_f);
   u8x8.drawString(0, 2, "Geiger-");
   u8x8.drawString(0, 3, " Counter");
   u8x8.drawString(0, 4, "Version:");
-  sprintf(line, "%d.%d.%d", VERSION_MAJOR, VERSION_MINOR, VERSION_PATCH);
   u8x8.drawString(0, 5, line);
   #else
+  // Display is only 4 lines by 16 characters; lines counting 0,2,4,6.
+  snprintf(line, 17, "%s", VERSION_STR);  // 16 chars + \0 termination
```
Comment (ThomasWaldmann) [inline multigeiger/multigeiger.ino]: using up to 16 chars here now.

File: multigeiger/multigeiger.ino [hunk 12]
```diff
@@ -751,28 +750,23 @@ int gen_charge_pulses(int max_charge_pulses) {
...
-  u8x8.println("==============");
-  sprintf(line, "V%d.%d.%d", VERSION_MAJOR, VERSION_MINOR, VERSION_PATCH);
-  u8x8.setCursor(7 - strlen(line) / 2, 4);
-  u8x8.print(line);
-  u8x8.setCursor(1, 6);
-  u8x8.print("Info:boehri.de");
+  u8x8.drawString(0, 0, "Geiger-Counter");
+  u8x8.drawString(0, 2, "==============");
+  u8x8.drawString(0, 4, line);
+  u8x8.drawString(0, 6, "Info:boehri.de");
```
Comment (ThomasWaldmann) [inline multigeiger/multigeiger.ino]: this is only 14 wide. we could add some chars:
- blanks left and right from Geiger-Counter
- 2x `=` more
- a blank between `Info:` and `boehri.de`

**Comment by ThomasWaldmann** [comment]
ok, read the u8x8 docs.

0,2,4,6 is correct for the double-height font (because the raster is always 8x8 and does not adjust to font size).

still unclear why it starts from 2 and clear does not work for the smaller display.

**Comment by ThomasWaldmann** [comment]
I reverted all display-related changes, they didn't really fit in here.

**Comment by ThomasWaldmann** [comment]
I'm merging this. As I removed all display-related changes anyway, this is now much simpler.

**Pull request merged**